p4-16/spec/P4-16-spec.mdk

```diff
@@ -1931,6 +1931,202 @@ described below.
+saturating arithmetic will clamp the value of the result to either the
+minimum or maximum value supported by the data type.
+
+All binary operations require both operands to have the same exact
+type, saturating, signedness, and width. Supplying operands of
+different types produces a compile-time error. No implicit casts
+(conversions) are inserted by the compiler to equalize the types. P4
+does not support any binary operations that combine saturating and
+modular arithmetic, except cast.
+
+The `saturating_int<W>` datatype supports the following operations:
+
+- Negation, denoted by unary `-`.
+- Unary plus, denoted by `+`. This operation behaves like a no-op.
+- Addition, denoted by `+`. Example:
+
+~ Begin P4Example
+saturating_int<W> a, b;
+int tmp = a + b;
+if (tmp < a.min)       // a.min = -2^(W-1)
+    a.min
+else if (tmp > a.max)  // a.max = 2^(W-1) - 1
+    a.max
+else
+    tmp
+~ End P4Example
+
+- Subtraction, denoted by `-`.
+- Multiplication, denoted by `*`. Result has the same width as the
+  operands. P4 architectures may impose additional
+  restrictions---e.g., they may only allow multiplication by a power
+  of two.  Multiplication semantics is equivalent to repeated
+  addition.
+- Comparison for equality and inequality, denoted `==` and `!=`
+  respectively. These operations produce a Boolean result.
+- Numeric comparisons, denoted by `<,<=,>,` and `>=`. These operations
+  produce a Boolean result.
+- Casting between saturating and non-saturating signed integers:
+  - Widening between saturating or saturating and non-saturating
+    signed: sign extend and extend the width of the type. The result
+    has the same value as before, with the new wider type (saturating
+    or non-saturating).
+  - Narrowing
+    - from saturating -> non-saturating: truncate the result to the
+      width of the desired type, while preserving the sign. If the
+      value was larger/smaller, the result will be truncated to the
+      maximum/minimum value on the narrower width with the appropriate
+      sign. Essentially, the same operation as a W-1 unsigned with the
+      preserved sign.
+    - from saturating/non-saturating -> saturating: truncate and clamp
+      the result to the width of the desired type, while preserving
+      the sign. If the value was larger/smaller, the result will be
+      clamped to the maximum/minimum on the narrower width with the
+      appropriate sign. Essentially, the same operation as a W-1
+      unsigned with the preserved sign.
+
+#### Using saturating arithmetic example
+
+We demonstrate the use of saturating arithmetic for preventing both
+underflow when decrementing `ttl`, and caping the value at maximum
+when using addition.
+
+~ Begin P4Example
+header ipv4_t {
+        bit<4>  version;
+        bit<4>  ihl;
+        bit<8>  diffserv;
+        bit<16> totalLen;
+        saturating_bit<16> identification;
+        bit<3>  flags : 3;
+        bit<13> fragOffset;
+        saturating_bit<8> ttl;
```

Comment:
This is the primary example of how NOT to use saturating_bit<>. Essentially the definition of a standard IP header now depends on how the programmer wants to to operate on it in the program, where as one would assume it should not depend on these details.